### ISSUE TITLE
twistedchecker-trunk-diff.sh 'git' command dangerous to use locally since it alternately deletes trunk

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8839]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#8839|
|Type|task|
|Created|2016-09-25 06:37:46Z|

the twistedchecker travis script does `git fetch origin trunk:refs/remotes/origin/trunk`; my refspec-fu is not quite up to the task of understanding this completely, but running that repeatedly does this:

```
$ git fetch origin trunk:refs/remotes/origin/trunk
From github:/twisted/twisted
 - [deleted]         (none)     -> origin/trunk
   (refs/remotes/origin/HEAD has become dangling)
$ git fetch origin trunk:refs/remotes/origin/trunk
From github:/twisted/twisted
 * [new branch]      trunk      -> origin/trunk
$ git fetch origin trunk:refs/remotes/origin/trunk
From github:/twisted/twisted
 - [deleted]         (none)     -> origin/trunk
   (refs/remotes/origin/HEAD has become dangling)
$ git fetch origin trunk:refs/remotes/origin/trunk
From github:/twisted/twisted
 * [new branch]      trunk      -> origin/trunk
```

which seems super bad.  In the environment in which it is supposed to run (travis) surely origin/trunk already exists?  and in other places the user can `git fetch`.  I suspect we should just delete the line.

<details><summary>Searchable metadata</summary>

```
trac-id__8839 8839
type__task task
reporter__glyph glyph
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1474785466029601 1474785466029601
changetime__1474842546262285 1474842546262285
version__None None
owner__Glyph__glyph_____ Glyph <glyph@...>

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Minor fix in review at https://github.com/twisted/twisted/pull/550

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Since runciter asked:

```
$ git fetch -v origin trunk:refs/remotes/origin/trunk
From github:/twisted/twisted
 - [deleted]         (none)     -> origin/trunk
   (refs/remotes/origin/HEAD has become dangling)
 = [up to date]      trunk      -> origin/trunk
$ git fetch -v origin trunk:refs/remotes/origin/trunk
From github:/twisted/twisted
 * [new branch]      trunk      -> origin/trunk
```

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_3"></a>|@markrwilliams commented|
|-|-|

OK to merge - review from PR:

I'm going to attempt to provide a gloss to this:

git fetch origin +refs/heads/trunk:refs/remotes/origin/trunk

...instructs git to fetch refs/heads/trunk from the remote named origin and store it locally in refs/remotes/origin/trunk. It appears that trunk in the source field of the refspec should be expanded to refs/heads/trunk. Maybe in this case it's not?

+ means "update the local reference even if it's not a fast forward." That seems a little worrisome to me, but it's the default behavior for fetch as configured by git remote add. You can confirm this by cloning a repository (since git clone implies git remote add) and checking .git/config:

```
[remote "origin"]
        <snip>
        fetch = +refs/heads/*:refs/remotes/origin/*
```

This seems OK, because my reading of the Git Book indicates that it's not different from what git fetch origin trunk:refs/remotes/origin/trunk should be anyway. So there should be no regression.

Sources:
https://git-scm.com/book/en/v2/Git-Internals-The-Refspec
https://www.atlassian.com/git/tutorials/refs-and-the-reflog/special-refs

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @glyph<br>@glyph set status to `closed`|
|-|-|

In changeset 620b84494a0d59309d27c1228bc3886771051a32
```
#!CommitTicketReference repository="" revision="620b84494a0d59309d27c1228bc3886771051a32"
Merge pull request #550 from twisted/8839-refspec

Author: glyph

Reviewer: markrwilliams

Fixes: #8839

Fix git command in txchecker tox environment to avoid breaking certain checkouts.
```
